Added support for postgres certificate authentication for native tls

linuxuser586 opened the pull request:
Fixes #1162  Adds support for postgres certificate authentication for native tls with openssl or rustls.

This allows using the uri format `postgres://postgres:password@localhost:5432/sqlx?sslmode=verify-full&sslrootcert=./ca.crt&sslcert=./client.crt&sslkey=./client.key` or the following environment variables

* `PGSSLROOTCERT`
* `PGSSLCERT`
* `PGSSLKEY`

This can be used for better security by disabling password login. It using the same patters used with psql.

Example 
```
# pg_hba.conf
# omit host to prevent password login
hostssl all all all cert
```

mehcode commented:
I'm thinking it might be best to allow a `native_tls::TlsConnector` or `rustls::Config` to be be passed to the connection options directly to allow anything.

linuxuser586 commented:
@mehcode I'm not sure if rusttls supports certificate authentication in the current implementation. I was trying to make the connection string match the way psql works. I had to use some openssl functions to handle the pkcs12 format. One way to get around this would be to do the pkcs12 conversion and specify that. I did that in my first test and it worked so seems possible for rustls. I will have to investigate more to see how to make it works with both tls supported options.

linuxuser586 commented:
@mehcode I finally found time to add support for certificate authentication using rustls. I updated the github workflows to test native tls using openssl, which is required for pkcs12 to identity transformation. I added a separate matrix for rustls without openssl since it does not need to use the pkcs12 format. I made openssl optional with native tls since the openssl library is only needed to support creating the identity from pkcs12 when using certificate authentication.

abonander commented:
@linuxuser586 sorry for the delay.

We had some recent changes to our Github Actions matrices that need to be rebased into this PR. Do you mind taking care of that?

linuxuser586 commented:
@abonander Done

abonander commented on sqlx-core/src/net/tls/mod.rs:
```diff
@@ -131,6 +143,26 @@ async fn configure_tls_connector(
+                    X509::from_pem(&cert_data),
+                ) {
+                    if let Ok(pkcs) =
+                        Pkcs12::builder().build(PASSWORD, PASSWORD, pkey.as_ref(), cert.as_ref())
```
If nothing else I would like to see comments added to this code explaining exactly _why_ it's doing this and what it needs to be better.

I also don't really like how the errors are discarded here.

abonander replied:
Honestly I would probably consider requiring the `openssl` crate for this, even as an optional feature, to be a non-starter. I would prefer to wait for the above PR to be merged, though it's not clear when that's going to happen.

abonander commented on sqlx-core/src/net/tls/mod.rs:
```diff
@@ -117,8 +121,16 @@ async fn configure_tls_connector(
     accept_invalid_certs: bool,
     accept_invalid_hostnames: bool,
     root_cert_path: Option<&CertificateInput>,
+    client_cert_path: Option<&CertificateInput>,
```
Can we fix these warnings? 

If the `native-tls` feature is enabled but not `openssl-native`, we should check if these are not `None` and if so, log a warning that these do nothing without the `openssl-native` feature.

abonander commented on sqlx-core/Cargo.toml:
```diff
@@ -93,11 +93,14 @@ _rt-actix = ["tokio-stream"]
 offline = ["serde", "either/serde"]
 
+# openssl
+openssl-native = [ "openssl" ]
```
I don't really see the reasoning behind giving this a different feature name. Why not just `openssl` like in the `sqlx` crate?

If you tried to define an `openssl` feature but got an error, it's because you don't need to explicitly define features for `optional = true` dependencies.

abonander reviewed (changes requested):
I was giving this a more in-depth review before merging it and came across a number of issues.

abonander commented:
Keep an eye on https://github.com/sfackler/rust-native-tls/pull/209, that looks promising.

emk commented:
Hello! Thank you for working on client certificates!

I'm the maintainer of https://github.com/emk/rust-musl-builder, a moderately popular tool for statically-linking Rust apps that use OpenSSL and `libpq` against `libmusl`. I'm trying to migrate away from that approach and encourage people to migrate to rustls, which is much easier to cross-compile and statically link.

Here is the code that I'm using for "pure Rust" TLS support, including some (as yet untested) client certificate support: https://github.com/dbcrossbar/dbcrossbar/pull/188/files. This uses `rustls` and `rustls-pemfile` to set up a single client certificate in a fashion much like that of `libpq`.

I would be happy to contribute any part of this code to help with this PR, or to prepare a draft PR based on this approach.

Google's hosted PostgreSQL servers actually require client certs to connect using TLS. So this is a potentially very useful feature.

abonander commented:
> Google's hosted PostgreSQL servers actually require client certs to connect using TLS. So this is a potentially very useful feature.

Which ones? Because we use PostgreSQL servers hosted by Google Cloud and it works fine without client certs. It just requires TLS in general.

larksoftware commented:
> Google's hosted PostgreSQL servers actually require client certs to connect using TLS. So this is a potentially very useful feature.

As a new user of GCP I was under this impression as well.  But after trying to configure sqlx with TLS and failing I did some more reading and found there is another method:

https://cloud.google.com/sql/docs/mysql/connect-admin-proxy
> Using the Cloud SQL Auth proxy is the recommended method for connecting to a Cloud SQL instance. The Cloud SQL Auth proxy:
>
>    Works with both public and private IP endpoints
    Validates connections using credentials for a user or service account
    Wraps the connection in a SSL/TLS layer that's authorized for a Cloud SQL instance

So using this auth proxy seems to be the recommended solution for connecting. I'm going to give this a try now and hope I can then use sqlx with my GCP postgres db.

linuxuser586 commented:
@abonander @larksoftware I experienced this very issue and is why I created the https://github.com/linuxuser586/sqlx/tree/postgres-native-tls-fixes fork. Unfortunately, my changes were never merged into the official version and I don't have the time to address the concerns to make it in to the official version; however, the forked version does meet my needs and may meet yours. It is easy to add the git version in rust and so for me, not using the official version is not an issue. I hope this helps you solve your problem.

abonander commented:
> I don't have the time to address the concerns to make it in to the official version

I guess we can close this PR then. My review comments still stand, as I would rather see proper support in `native-tls` than the round-tripping through OpenSSL here.

Note to anyone using this fork, it doesn't properly bubble up errors if there's something wrong with the certificate, it'll just panic: https://github.com/launchbadge/sqlx/pull/1166#discussion_r714298936

abonander commented:
Code quality aside, the solution for `rustls` _was_ acceptable so if anyone wants to open a PR based on this one with just the `rustls` changes and address the nits, we'd gladly merge it.
